ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs

```diff
@@ -211,25 +216,25 @@ openDBImpl :: (HasCallStack, IOLike m, Ord blockId, NoUnexpectedThunks blockId)
            -> ThrowCantCatch (VolatileDBError blockId) (STM m)
            -> Parser e m blockId
            -> Int -- ^ @maxBlocksPerFile@
+           -> ResourceRegistry m
            -> m (VolatileDBEnv m blockId)
-openDBImpl hasFS@HasFS{..} err errSTM parser maxBlocksPerFile =
+openDBImpl hasFS@HasFS{..} err errSTM parser maxBlocksPerFile registry =
     if maxBlocksPerFile <= 0
     then EH.throwError err $ UserError . InvalidArgumentsError $
       "maxBlocksPerFile should be positive"
     else do
-      st <- mkInternalStateDB hasFS err parser maxBlocksPerFile
+      st <- mkInternalStateDB hasFS err parser maxBlocksPerFile registry
       stVar <- newMVar $ VolatileDbOpen st
-      return $ VolatileDBEnv hasFS err errSTM stVar maxBlocksPerFile parser
+      return $
+        VolatileDBEnv hasFS err errSTM stVar maxBlocksPerFile parser registry
 
 closeDBImpl :: IOLike m
             => VolatileDBEnv m blockId
             -> m ()
-closeDBImpl VolatileDBEnv{..} = do
-    mbInternalState <- swapMVar _dbInternalState VolatileDbClosed
-    case mbInternalState of
-      VolatileDbClosed -> return ()
-      VolatileDbOpen InternalState{..} ->
-        wrapFsError hasFsErr _dbErr $ hClose _currentWriteHandle
+closeDBImpl VolatileDBEnv{..} =
+    bracket (takeMVar _dbInternalState)
+            (\_ -> putMVar _dbInternalState VolatileDbClosed)
+            (\_ -> unsafeReleaseAll _registry)
```

Comment:
Note the `wrapFsError` that we had, where does it live now? It wrapped `FsErrror`s in `UnexpectedError.FileSystemError`.